common/zeroconf/zeroconf.cpp

```diff
@@ -0,0 +1,130 @@
+
+namespace Aseba
+{
+	using namespace Dashel;
+
+	//! Register and announce a target described by a name and a port
+	void Zeroconf::registerPort(const string name, const unsigned port, const string txt)
+	{
+        ZeroconfService z;
+        uint16_t len = txt.size();
+        const char* record = txt.c_str();
+        DNSServiceErrorType err = DNSServiceRegister(&z.serviceref,
+                                                     kDNSServiceFlagsDefault,
+                                                     0, // default all interfaces
+                                                     name.c_str(),
+                                                     "_aseba._tcp",
+                                                     NULL, // use default domain
+                                                     NULL, // use this host name
+                                                     htons(port),
+                                                     len, // TXT length
+                                                     record, // TXT record
+                                                     cb_Register,
+                                                     &z);
+        if (err != kDNSServiceErr_NoError)
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegister: error %0").arg(err));
+        else
+        {
+            DNSServiceErrorType err = DNSServiceProcessResult(z.serviceref); // block until daemon replies
+            if (err != kDNSServiceErr_NoError)
+                throw runtime_error(FormatableString("Zeroconf DNSServiceProcessResult: error %0").arg(err));
+            else
+            {
+                services[z.name] = z; // copy
+                z.serviceref = 0; // because now owned by services
+            }
+        }
+	}
+
+    //! Register and announce a target described by an existing Dashel stream
+    void Zeroconf::registerStream(const std::string name, const Dashel::Stream* stream, const string txt)
+    {
+        string service_name = name;
+        unsigned port = atoi(stream->getTargetParameter("port").c_str());
+        registerPort(name,port,txt);
+    }
+
+    //! Asynchronously update the set of known targets
+    void Zeroconf::browseForTargets()
+    {
+    }
+
+    //! callback to update ZeroconfService record with results of registration
+    void DNSSD_API Zeroconf::cb_Register(DNSServiceRef sdRef,
+                               DNSServiceFlags flags,
+                               DNSServiceErrorType errorCode,
+                               const char *name,
+                               const char *regtype,
+                               const char *domain,
+                               void *context)
+    {
+        if (errorCode != kDNSServiceErr_NoError)
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegisterReply: error %0").arg(errorCode));
+        else
+        {
+            ZeroconfService *zref = (ZeroconfService *)context;
```

Comment:
Void isn't a class. `DNSServiceRegister` sends an "application context pointer" of our choosing to the callback, it could be NULL, but here I use it to identify the `ZeroconfService` instance that needs to be updated with the name that was actually chosen by the OS—the default behavior of `DNSServiceRegister` is to add a numerical suffix to a name when a conflict occurs e.g. "dummynode-0 (2)".

Alternative 1) Disallow renaming: an attempt to register a Zeroconf name that has already been used will throw an exception. This is probably not a problem, since the proposal in #543 includes the node id in the name, e.g. "Thymio Wireless 5036".

Alternative 2) Maintain a `DNSServiceRef` to `ZeroconfService*` map in the `Zeroconf` instance.

(The updates to `regtype` and `domain` should be removed because these values don't change.)